src/cookbooks/clients_last_seen_bits.md

```diff
@@ -0,0 +1,375 @@
+WHERE
+  submission_date = '2020-01-28'
+```
+
+### Retention using activity date
+
+GUD's canonical retention definitions are all based on ping submission dates rather
+than logical activity dates taken from client-provided timestamps, but there is
+interest in using client timestamps particular for `n`-day retention calculations
+for mobile products.
+
+Let's consider Firefox Preview which sends telemetry via Glean. The
+`org_mozilla_fenix.baseline_clients_last_seen` table includes two bit patterns
+that encode client timestamps: `days_seen_session_start_bits` and
+`days_seen_session_end_bits`. This table is still populated once per day based
+on pings received over the previous day, but some of those pings will reflect
+sessions that started on previous days. This introduces some new complexity
+into retention calculations, because we'll always be undercounting clients
+if we have our retention window end on `submission_date`.
+
+When using activity date, it may be desirable to build in a few days of buffer
+to ensure we are considering late-arriving pings. For example, if we wanted
+to calculate 3-day retention but allow 2 days of cushion for late-arriving
+pings, we would need to use an offset of 5 days from `submission_date`:
+
+```
+DECLARE n, cushion_days, offset_to_day_0 INT64;
+SET n = 3;
+SET cushion_days = 2;
+SET offset_to_day_0 = n + cushion_days;
+
+WITH base AS (
+  SELECT
+    *,
+    udf.bits28_range(days_seen_session_start_bits, offset_to_day_0, 0, 1) AS seen_on_day_0,
+    udf.bits28_range(days_seen_session_start_bits, offset_to_day_0, 1, n - 1) AS seen_after_day_0
+  FROM
+    org_mozilla_fenix.baseline_clients_last_seen )
+SELECT
+  DATE_SUB(submission_date, INTERVAL offset_to_day_0 DAYS) AS metric_date,
+  SAFE_DIVIDE(
+    COUNTIF(seen_on_day_0 AND seen_after_day_0),
+    COUNTIF(seen_on_day_0)
+  ) AS retention_n_day
+FROM
+  base
+GROUP BY
+  metric_date
+WHERE
+  submission_date = '2020-01-28'
+```
+
+## UDF Reference
+
+### bits28_to_string
+
+Convert an INT64 field into a 28-character string representing the individual bits.
+
+```
+bits28_to_string(bits INT64)
+
+SELECT udf.bits28_to_string(18)
+>> 0000000000000000000000010010
+```
+
+### bits28_from_string
+
+Convert a string representing individual bits into an INT64.
+
+```
+bits28_from_string(bits STRING)
+
+SELECT udf.bits28_from_string('10010')
+>> 18
+```
+
+### bits28_to_dates
+
+Convert a bit pattern into an array of the dates is represents.
+
+```
+bits28_to_dates(bits INT64, end_date DATE)
+
+SELECT udf.bits28_to_dates(18, '2020-01-28')
+>> ['2020-01-24', '2020-01-27']
+```
+
+### bits28_days_since_seen
+
+Return the position of the rightmost set bit in an INT64 bit pattern.
+
+```
+bits28_days_since_seen(bits INT64)
+
+SELECT bits28_days_since_seen(18)
+>> 1
+```
+
+
+### bits28_range
+
+Return an INT64 representing a range of bits from a source bit pattern.
+
+```
+bits28_range(bits INT64, offset_to_day_0 INT64, start_bit INT64, number_of_bits INT64)
```

Comment:
This ` bits28_range` function is the trickiest and has the highest margin for error. It has a gaggle of params, all of which are ints, so ripe for getting the order wrong, etc.

This function also kind of does a lot. We could instead express this as two separate functions, one that trims off bits from the front and one that trims off bits from the end, something like:

```
-- Get bits 7 though 13.
bits28_take_first(bits28_take_last(bits, 14), 7)
```

Reply:
I'm thinking that potentially a more simplistic function would actually be easier to reason about if we provide sufficient documentation on usage.

We could have `bits28_range(bits, start_index, number_of_bits)` where the index is a negative number that indicates a number of bits from the end of the pattern. So dealing with a 2-week window could look like:

```
SELECT
  DATE_SUB(submission_date, INTERVAL 13 DAY) AS metric_date,
  BIT_COUNT(udf.bits28_range(days_seen_bits, -13 + 0, 1)) > 0 AS active_on_metric_date,
  BIT_COUNT(udf.bits28_range(days_seen_bits, -13 + 0, 7)) > 0 AS active_in_week_0,
  BIT_COUNT(udf.bits28_range(days_seen_bits, -13 + 1, 6)) > 0 AS active_in_week_0_after_metric_date,
  BIT_COUNT(udf.bits28_range(days_seen_bits, -13 + 7, 7)) > 0 AS active_in_week_1,
```